### ISSUE TITLE
In XR editor functionalities

### ISSUE THREAD

**Issue by vincentfretin** [c-frame/aframe-editor#5]
@Utopiah is interested in a VR editor mode, components for that may leave in a separate repo as a plugin for aframe-editor (to be defined). If there are any change to do in this repo that would make easier to support that, we can do it.
Some notes on his repo:
https://git.benetou.fr/utopiah/text-code-xr-engine/issues/99

**Comment by Utopiah** [comment]
A specific constraint I do have, which is a personal design preference, is to be "no-build" in order to iterative faster. This is particularly important for as I give workshops to non professional developers. For such a crowd, but even for myself I admit, being to "just" edit the content and see a change without going first through intermediary steps (i.e installing node, npm, going through the command line, etc) is fundamental.

If going through a plugin solution, unless it's entirely impossible, I would much rather be able to avoid any build step.

**Comment by Utopiah** [comment]
Some progress on that, namely cloning primitives https://x.com/utopiah/status/1842195658637464037 and changing their color https://x.com/utopiah/status/1842198368438669389

**Comment by vincentfretin** [comment]
@Utopiah For your project https://nlnet.nl/project/ImmersiveXP/ if your XR interaction components to move elements and edit components on it rely on the commands documented here https://github.com/c-frame/aframe-editor/blob/dev/docs/commands.md instead of doing a createElement/setAttribute yourself, you will have some features out of the box, like undo/redo, incremental auto-save, and collaborative editing once I implement them in aframe-editor.
I could add an option in aframe-editor to not render the React UI so you can just use the core to have the commands. If you want to consume aframe-editor as a library via cdn, I could expose Events.off/Events.on if you need to listen on some events.

And inversely we could reuse your XR interaction components in aframe-editor.

As @kfarr mentioned in the linked issue, the three.js editor https://threejs.org/editor added rendering the transform controls and sidebar in XR
https://twitter.com/mrdoob/status/1758531748483485842
That's using HTMLMesh in https://github.com/mrdoob/three.js/blob/dev/editor/js/Viewport.XR.js
We may need to update the TransformControls we use in aframe-editor to have that working in XR.

Also see the recent changes in HTMLMesh for XR:
- https://github.com/AdaRoseCannon/aframe-htmlmesh/pull/25
- https://github.com/AdaRoseCannon/aframe-htmlmesh/pull/24

So we could also render the components sidebar with aframe-htmlmesh, but probably do something more specific for XR, showing a flat UI with position/rotation/scale may not be needed, maybe add some UI on the wrist corresponding to the selected element.

**Comment by foobar404** [comment]
@Utopiah has there been any progress on the XR editor mode, and is the code public somewhere?

**Comment by Utopiah** [comment]
@foobar404 yes there are been progress and the code is public but it's quite a mess.

You can see : 
- videos at https://video.benetou.fr/videos/recently-added?s=1
- code https://git.benetou.fr/explore/repos in particular
  -  [biggu_s_gate](https://git.benetou.fr/utopiah/biggu_s_gate)
  - [spasca-fot-sloan-q1](https://git.benetou.fr/utopiah/spasca-fot-sloan-q1)
  - [spasca-fot-sloan-q2](https://git.benetou.fr/utopiah/spasca-fot-sloan-q2)
  - [spasca-fot-sloan-q3](https://git.benetou.fr/utopiah/spasca-fot-sloan-q3)
  - [spasca-fot-sloan-q4](https://git.benetou.fr/utopiah/spasca-fot-sloan-q4)
  - [text-code-xr-engine](https://git.benetou.fr/utopiah/text-code-xr-engine)

... but it's more like disparate individual demos than a structured set.

If you have some specific need in mind happy to point you to something more specific, go through the code or collect everything required then make it a dedicated component.

We can discuss that here, chat via Discord or my Jitsi Meet server synchronously.

**Comment by foobar404** [comment]
@Utopiah The videos are interesting. I see you do a lot of work with hand tracking, something i currently know little about. Currently im working on a fork of the aframe-inspector, with an XR mode built in https://github.com/foobar404/aframe-engine . Its still a work in progress, my goal with it is to have a XR version that is as feature complete as the 2d version, ideally without just dropping in 2d panels to fill in the functionality. Your text-code-xr-engine repo could be helpful for when I start working on scripting in XR. Ideally, the XR mode will also have a hand tracking mode, currently only controllers are supported. Changes made in XR will use the same event system that the 2d editor uses so they can be synced using a modified version of the aframe-watcher. If this sounds interesting, feel free to check out the repo and report bugs / request features.
